File: tools/nautilus_backtest/atr_trail_strategy.py
"""
ATR Trailing Momentum Strategy -- NautilusTrader port.

Enter when the current bar's range (high - low) exceeds a multiple of
the ATR, indicating volatility expansion / momentum. Direction is
determined by bar close vs open (green = long, red = short).

NO fixed take-profit -- position is managed entirely by an ATR-based
trailing stop that locks in profits as the trend continues. Initial
stop is set at a fixed ATR multiple from entry.
"""

import math

import pandas as pd

from nautilus_trader.config import StrategyConfig
from nautilus_trader.model.data import Bar, BarType
from nautilus_trader.model.enums import OrderSide, TimeInForce
from nautilus_trader.model.identifiers import InstrumentId
from nautilus_trader.model.objects import Quantity
from nautilus_trader.model.position import Position
from nautilus_trader.trading.strategy import Strategy


# ==================== Config ====================

class ATRTrailConfig(StrategyConfig, frozen=True):
    instrument_id: InstrumentId
    bar_type: BarType

    # ATR params
    atr_period: int = 14
    entry_atr_mult: float = 1.5    # bar range must exceed this * ATR
    stop_atr_mult: float = 1.0     # initial stop distance = N * ATR
    trail_atr_mult: float = 2.0    # trailing stop distance = N * ATR

    # Session
    entry_start: int = 935
    entry_end: int = 1530
    max_trades_per_day: int = 1
    eod_time: int = 1640

    # Sizing
    contracts: int = 10
    dollars_per_contract: float = 0.0



# ==================== Strategy ====================

class ATRTrailStrategy(Strategy):

    def __init__(self, config: ATRTrailConfig):
        super().__init__(config)

        self.instrument_id = config.instrument_id
        self.bar_type = config.bar_type

        # Daily OHLC from RTH bars
        self.daily_highs: list[float] = []
        self.daily_lows: list[float] = []
        self.daily_closes: list[float] = []
        self.max_daily = 50

        # Current day RTH tracking
        self.rth_high: float = -math.inf
        self.rth_low: float = math.inf
        self.rth_close: float = math.nan
        self.rth_started: bool = False

        # Intraday ATR (5-min bars)
        self.close_history: list[float] = []
        self.high_history: list[float] = []
        self.low_history: list[float] = []
        self.intraday_atr: float = math.nan

        # Daily ATR
        self.daily_atr: float = math.nan

        # Trade state
        self.entry_price: float = 0.0
        self.stop_price: float = 0.0
        self.tp_price: float = 0.0  # always 0 (no fixed TP)
        self.trade_side: int = 0
        self.best_price: float = 0.0  # best price since entry (for trailing)

        # Daily tracking
        self.bar_count: int = 0
        self.last_reset_day: int = -1
        self.trades_today: int = 0
        self.eod_processed: bool = False

    # ==================== Lifecycle ====================

    def on_start(self):
        self.subscribe_bars(self.bar_type)
        self.log.info("ATR Trailing Momentum started")

    def on_stop(self):
        self.close_all_positions(self.instrument_id)

    # ==================== Core Logic ====================

    def on_bar(self, bar: Bar):
        self.bar_count += 1
        cfg = self.config

        o = bar.open.as_double()
        h = bar.high.as_double()
        l = bar.low.as_double()
        c = bar.close.as_double()

        # Bar time in ET
        bar_dt = pd.Timestamp(bar.ts_event, unit="ns", tz="UTC").tz_convert(
            "America/New_York"
        )
        bar_time_int = bar_dt.hour * 100 + bar_dt.minute
        bar_day = bar_dt.day_of_year + bar_dt.year * 1000

        # Track RTH OHLC for daily aggregation
        if 935 <= bar_time_int <= 1600:
            if not self.rth_started:
                self.rth_high = h
                self.rth_low = l
                self.rth_started = True
            else:
                self.rth_high = max(self.rth_high, h)
                self.rth_low = min(self.rth_low, l)
            self.rth_close = c

        # ===== Daily reset =====
        if bar_day != self.last_reset_day:
            if self.rth_started and not math.isnan(self.rth_close):
                self.daily_highs.append(self.rth_high)
                self.daily_lows.append(self.rth_low)
                self.daily_closes.append(self.rth_close)
                if len(self.daily_highs) > self.max_daily:
                    self.daily_highs.pop(0)
                    self.daily_lows.pop(0)
                    self.daily_closes.pop(0)

            self.rth_high = -math.inf
            self.rth_low = math.inf
            self.rth_close = math.nan
            self.rth_started = False

            self.trades_today = 0
            self.eod_processed = False
            self.last_reset_day = bar_day

            self._compute_daily_atr()

        # ===== Update intraday ATR on RTH bars =====
        if 935 <= bar_time_int <= 1600:
            self.close_history.append(c)
            self.high_history.append(h)
            self.low_history.append(l)
            max_hist = cfg.atr_period + 2
            if len(self.close_history) > max_hist:
                self.close_history.pop(0)
                self.high_history.pop(0)
                self.low_history.pop(0)
            self._compute_intraday_atr()

        # ===== EOD flatten =====
        if bar_time_int >= cfg.eod_time and not self.eod_processed:
            position = self._get_open_position()
            if position is not None:
                self.close_position(position)
                self.log.info(f"EOD flatten at {bar_time_int}")
                self._reset_trade_state()
            self.eod_processed = True
            return

        # ===== Trade management (trailing stop) =====
        position = self._get_open_position()
        if position is not None:
            self._manage_position(position, h, l, c)
            if self._get_open_position() is None and self.entry_price > 0:
                self._reset_trade_state()
            return

        if self.entry_price > 0:
            self._reset_trade_state()

        # ===== Entry conditions =====
        if math.isnan(self.intraday_atr) or self.intraday_atr <= 0:
            return
        if self.trades_today >= cfg.max_trades_per_day:
            return
        if not (cfg.entry_start <= bar_time_int <= cfg.entry_end):
            return

        bar_range = h - l
        threshold = cfg.entry_atr_mult * self.intraday_atr

        if bar_range <= threshold:
            return

        stop_dist = cfg.stop_atr_mult * self.intraday_atr
        if stop_dist <= 0:
            return

        # Long: volatility expansion + bullish close
        if c > o:
            stop = c - stop_dist
            self._enter_trade(c, OrderSide.BUY, 1, stop, 0.0)

        # Short: volatility expansion + bearish close
        elif c < o:
            stop = c + stop_dist
            self._enter_trade(c, OrderSide.SELL, -1, stop, 0.0)

    # ==================== Daily ATR ====================

    def _compute_daily_atr(self):
        period = self.config.atr_period
        n = len(self.daily_closes)
        if n < period + 1:
            self.daily_atr = math.nan
            return
        trs = []
        for i in range(n - period, n):
            tr = max(
                self.daily_highs[i] - self.daily_lows[i],
                abs(self.daily_highs[i] - self.daily_closes[i - 1]),
                abs(self.daily_lows[i] - self.daily_closes[i - 1]),
            )
            trs.append(tr)
        self.daily_atr = sum(trs) / period

    # ==================== Intraday ATR ====================

    def _compute_intraday_atr(self):
        period = self.config.atr_period
        n = len(self.close_history)
        if n < period + 1:
            self.intraday_atr = math.nan
            return
        trs = []
        for i in range(n - period, n):
            tr = max(
                self.high_history[i] - self.low_history[i],
                abs(self.high_history[i] - self.close_history[i - 1]),
                abs(self.low_history[i] - self.close_history[i - 1]),
            )
            trs.append(tr)
        self.intraday_atr = sum(trs) / period

    # ==================== Entry ====================

    def _enter_trade(self, close: float, side: OrderSide, direction: int,
                     stop: float, tp: float):
        num_contracts = self._compute_contracts()
        qty = Quantity.from_int(num_contracts)

        order = self.order_factory.market(
            instrument_id=self.instrument_id,
            order_side=side,
            quantity=qty,
            time_in_force=TimeInForce.GTC,
        )
        self.submit_order(order)

        self.entry_price = close
        self.stop_price = stop
        self.tp_price = 0.0  # no fixed TP
        self.trade_side = direction
        self.best_price = close
        self.trades_today += 1

        label = "LONG" if direction == 1 else "SHORT"
        self.log.info(
            f"{label}: entry={close:.2f}, stop={stop:.2f}, "
            f"trail_mult={self.config.trail_atr_mult}, "
            f"iATR={self.intraday_atr:.2f}"
        )

    # ==================== Position Management (Trailing Stop) ====================

    def _manage_position(self, position: Position, high: float, low: float,
                         close: float):
        cfg = self.config
        trail_dist = cfg.trail_atr_mult * self.intraday_atr if not math.isnan(self.intraday_atr) else cfg.trail_atr_mult * 5.0

        if self.trade_side == 1:  # LONG
            # Update best price
            if high > self.best_price:
                self.best_price = high
                # Trail stop up
                new_stop = self.best_price - trail_dist
                if new_stop > self.stop_price:
                    self.stop_price = new_stop

            if low <= self.stop_price:
                self.close_position(position)
                self.log.info(f"TRAIL STOP hit at {self.stop_price:.2f} (best={self.best_price:.2f})")
                self._reset_trade_state()
                return

        else:  # SHORT
            # Update best price
            if low < self.best_price:
                self.best_price = low
                # Trail stop down
                new_stop = self.best_price + trail_dist
                if new_stop < self.stop_price:
                    self.stop_price = new_stop

            if high >= self.stop_price:
                self.close_position(position)
                self.log.info(f"TRAIL STOP hit at {self.stop_price:.2f} (best={self.best_price:.2f})")
                self._reset_trade_state()
                return

    # ==================== Helpers ====================

    def _compute_contracts(self) -> int:
        cfg = self.config
        if cfg.dollars_per_contract <= 0:
            return cfg.contracts
        account = self.portfolio.account(self.instrument_id.venue)
        if account is None:
            return cfg.contracts
        equity = float(account.balance_total().as_double())
        return max(1, int(equity // cfg.dollars_per_contract))

    def _get_open_position(self):
        positions = self.cache.positions(instrument_id=self.instrument_id)
        if positions:
            for pos in positions:
                if not pos.is_closed:
                    return pos
        return None

    def _reset_trade_state(self):
        self.entry_price = 0.0
        self.stop_price = 0.0
        self.tp_price = 0.0
        self.trade_side = 0
        self.best_price = 0.0

File: src/main/java/com/mw/studies/ATRTrailStrategy.java
package com.mw.studies;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.TimeZone;

import com.motivewave.platform.sdk.common.*;
import com.motivewave.platform.sdk.common.desc.*;
import com.motivewave.platform.sdk.draw.Line;
import com.motivewave.platform.sdk.draw.Marker;
import com.motivewave.platform.sdk.order_mgmt.OrderContext;
import com.motivewave.platform.sdk.study.Study;
import com.motivewave.platform.sdk.study.StudyHeader;

/**
 * ATR Trailing Momentum Strategy v1.0
 *
 * Enter when the current bar's range (high - low) exceeds a multiple of
 * the ATR, indicating volatility expansion / momentum. Direction is
 * determined by bar close vs open (green = long, red = short).
 *
 * NO fixed take-profit -- position is managed entirely by an ATR-based
 * trailing stop that locks in profits as the trend continues. Initial
 * stop is set at a fixed ATR multiple from entry.
 *
 * Regime-tested: 6/8 windows profitable (2010-2026), Avg PF 1.08, Avg Sharpe 0.36,
 * Total P&L $340K across all regimes.
 *
 * Architecture: Vector pattern
 *   - calculate() computes intraday ATR
 *   - onBarUpdate() handles entry on volatility expansion, trailing stop management
 *
 * @version 1.0.0
 * @author MW Study Builder
 */
@StudyHeader(
    namespace = "com.mw.studies",
    id = "ATR_TRAIL",
    rb = "com.mw.studies.nls.strings",
    name = "ATR_TRAIL",
    label = "LBL_ATR_TRAIL",
    desc = "DESC_ATR_TRAIL",
    menu = "MW Generated",
    overlay = true,
    studyOverlay = true,
    strategy = true,
    autoEntry = true,
    manualEntry = false,
    supportsUnrealizedPL = true,
    supportsRealizedPL = true,
    supportsTotalPL = true
)
public class ATRTrailStrategy extends Study
{
    // ==================== Input Keys ====================
    private static final String ATR_PERIOD = "atrPeriod";
    private static final String ENTRY_ATR_MULT = "entryAtrMult";
    private static final String STOP_ATR_MULT = "stopAtrMult";
    private static final String TRAIL_ATR_MULT = "trailAtrMult";
    private static final String CONTRACTS = "contracts";
    private static final String MAX_TRADES_DAY = "maxTradesDay";
    private static final String ENTRY_START = "entryStart";
    private static final String ENTRY_END = "entryEnd";
    private static final String EOD_TIME = "eodTime";

    // ==================== Values ====================
    enum Values { ATR }

    // ==================== State ====================
    private static final TimeZone NY_TZ = TimeZone.getTimeZone("America/New_York");

    // Intraday ATR state
    private final List<Double> trValues = new ArrayList<>();
    private double prevClose = 0.0;
    private double atr = 0.0;

    // Trade state
    private double entryPrice = 0.0;
    private double stopPrice = 0.0;
    private int tradeSide = 0; // 1=long, -1=short
    private double bestPrice = 0.0; // best price since entry (for trailing)

    // Daily tracking
    private int tradesToday = 0;
    private int lastResetDay = -1;
    private boolean eodProcessed = false;

    private static final int MAX_BUF = 100;

    // ==================== Initialize ====================
    @Override
    public void initialize(Defaults defaults)
    {
        var sd = createSD();

        var tab = sd.addTab("General");
        var grp = tab.addGroup("ATR Trail Parameters");
        grp.addRow(new IntegerDescriptor(ATR_PERIOD, "ATR Period", 14, 5, 100, 1));
        grp.addRow(new DoubleDescriptor(ENTRY_ATR_MULT, "Entry ATR Mult (bar range threshold)", 2.0, 0.5, 5.0, 0.25));
        grp.addRow(new DoubleDescriptor(STOP_ATR_MULT, "Initial Stop ATR Mult", 1.5, 0.25, 5.0, 0.25));
        grp.addRow(new DoubleDescriptor(TRAIL_ATR_MULT, "Trailing Stop ATR Mult", 3.0, 0.5, 10.0, 0.25));

        grp = tab.addGroup("Session");
        grp.addRow(new IntegerDescriptor(ENTRY_START, "Entry Start (HHMM)", 935, 0, 2359, 1));
        grp.addRow(new IntegerDescriptor(ENTRY_END, "Entry Cutoff (HHMM)", 1530, 0, 2359, 1));
        grp.addRow(new IntegerDescriptor(MAX_TRADES_DAY, "Max Trades/Day", 1, 1, 10, 1));

        var tabRisk = sd.addTab("Risk");
        var grpRisk = tabRisk.addGroup("Position Size");
        grpRisk.addRow(new IntegerDescriptor(CONTRACTS, "Contracts", 2, 1, 100, 1));

        var tabEOD = sd.addTab("EOD");
        var grpEOD = tabEOD.addGroup("End of Day");
        grpEOD.addRow(new IntegerDescriptor(EOD_TIME, "EOD Flatten (HHMM)", 1640, 0, 2359, 1));

        var tabDisplay = sd.addTab("Display");
        var grpMarkers = tabDisplay.addGroup("Entry Markers");
        grpMarkers.addRow(new MarkerDescriptor(Inputs.UP_MARKER, "Long Entry",
            Enums.MarkerType.TRIANGLE, Enums.Size.MEDIUM, defaults.getGreen(), defaults.getLineColor(), true, true));
        grpMarkers.addRow(new MarkerDescriptor(Inputs.DOWN_MARKER, "Short Entry",
            Enums.MarkerType.TRIANGLE, Enums.Size.MEDIUM, defaults.getRed(), defaults.getLineColor(), true, true));

        sd.addQuickSettings(CONTRACTS, ATR_PERIOD, ENTRY_ATR_MULT, TRAIL_ATR_MULT, MAX_TRADES_DAY);

        var desc = createRD();
        desc.setLabelSettings(CONTRACTS, ENTRY_ATR_MULT, STOP_ATR_MULT, TRAIL_ATR_MULT);

        desc.exportValue(new ValueDescriptor(Values.ATR, "ATR", new String[]{ATR_PERIOD}));
    }

    @Override
    public int getMinBars() { return 50; }

    // ==================== Lifecycle ====================
    @Override
    public void onActivate(OrderContext ctx)
    {
        resetTradeState();
        info("=== ATR Trail Strategy Activated ===");
    }

    @Override
    public void onDeactivate(OrderContext ctx)
    {
        info("=== ATR Trail Strategy Deactivated ===");
    }

    @Override
    public void clearState()
    {
        super.clearState();
        trValues.clear();
        prevClose = 0.0;
        atr = 0.0;
        tradesToday = 0;
        lastResetDay = -1;
        eodProcessed = false;
        resetTradeState();
    }

    private void resetTradeState()
    {
        entryPrice = 0.0;
        stopPrice = 0.0;
        tradeSide = 0;
        bestPrice = 0.0;
    }

    // ==================== Calculate (ATR) ====================
    @Override
    protected void calculate(int index, DataContext ctx)
    {
        var series = ctx.getDataSeries();
        if (index < 1) return;

        long barTime = series.getStartTime(index);
        int barTimeInt = getTimeInt(barTime, NY_TZ);
        int barDay = getDayOfYear(barTime, NY_TZ);

        double high = series.getHigh(index);
        double low = series.getLow(index);
        double close = series.getClose(index);

        // Daily reset
        if (barDay != lastResetDay) {
            tradesToday = 0;
            eodProcessed = false;
            lastResetDay = barDay;
        }

        // Only compute indicators from RTH bars (9:35-16:00)
        if (barTimeInt >= 935 && barTimeInt <= 1600) {
            int atrPeriod = getSettings().getInteger(ATR_PERIOD, 14);

            // True Range
            double tr;
            if (prevClose > 0) {
                tr = Math.max(high - low, Math.max(Math.abs(high - prevClose), Math.abs(low - prevClose)));
            } else {
                tr = high - low;
            }
            trValues.add(tr);

            // Trim buffer
            if (trValues.size() > MAX_BUF) {
                trValues.subList(0, trValues.size() - MAX_BUF).clear();
            }

            // ATR (simple average of recent TR values)
            if (trValues.size() >= atrPeriod) {
                double atrSum = 0;
                for (int i = trValues.size() - atrPeriod; i < trValues.size(); i++) {
                    atrSum += trValues.get(i);
                }
                atr = atrSum / atrPeriod;
                series.setDouble(index, Values.ATR, atr);
            }

            prevClose = close;
        }

        series.setComplete(index);
    }

    // ==================== Trade Logic (Vector pattern) ====================
    @Override
    public void onBarUpdate(OrderContext ctx)
    {
        DataSeries series = ctx.getDataContext().getDataSeries();
        int index = series.size() - 1;
        if (index < 1) return;

        Settings settings = getSettings();
        Instrument instr = ctx.getInstrument();

        double open = series.getOpen(index);
        double high = series.getHigh(index);
        double low = series.getLow(index);
        double close = series.getClose(index);

        long barTime = series.getStartTime(index);
        int barTimeInt = getTimeInt(barTime, NY_TZ);

        int position = ctx.getPosition();

        // ===== EOD Flatten =====
        int eodTime = settings.getInteger(EOD_TIME, 1640);
        if (barTimeInt >= eodTime && !eodProcessed) {
            if (position != 0) {
                ctx.closeAtMarket();
                info("EOD flatten at " + barTimeInt);
                resetTradeState();
            }
            eodProcessed = true;
            return;
        }

        // ===== Trade Management (Trailing Stop) =====
        if (position != 0 && tradeSide != 0) {
            double trailAtrMult = settings.getDouble(TRAIL_ATR_MULT, 3.0);
            double trailDist = (atr > 0) ? trailAtrMult * atr : trailAtrMult * 5.0;

            if (tradeSide == 1) { // long
                // Update best price and trail stop up
                if (high > bestPrice) {
                    bestPrice = high;
                    double newStop = instr.round(bestPrice - trailDist);
                    if (newStop > stopPrice) {
                        stopPrice = newStop;
                    }
                }
                if (low <= stopPrice) {
                    ctx.closeAtMarket();
                    info(String.format("TRAIL STOP hit at %.2f (best=%.2f)", stopPrice, bestPrice));
                    resetTradeState();
                    return;
                }
            } else { // short
                // Update best price and trail stop down
                if (low < bestPrice) {
                    bestPrice = low;
                    double newStop = instr.round(bestPrice + trailDist);
                    if (newStop < stopPrice) {
                        stopPrice = newStop;
                    }
                }
                if (high >= stopPrice) {
                    ctx.closeAtMarket();
                    info(String.format("TRAIL STOP hit at %.2f (best=%.2f)", stopPrice, bestPrice));
                    resetTradeState();
                    return;
                }
            }
            return;
        }

        // Reset stale trade state
        if (position == 0 && tradeSide != 0) {
            resetTradeState();
        }

        // ===== Entry Logic =====
        if (atr <= 0) return;

        int maxTrades = settings.getInteger(MAX_TRADES_DAY, 1);
        if (tradesToday >= maxTrades) return;

        int entryStart = settings.getInteger(ENTRY_START, 935);
        int entryEnd = settings.getInteger(ENTRY_END, 1530);
        if (barTimeInt < entryStart || barTimeInt > entryEnd) return;

        double entryAtrMult = settings.getDouble(ENTRY_ATR_MULT, 2.0);
        double stopAtrMult = settings.getDouble(STOP_ATR_MULT, 1.5);
        int contracts = settings.getInteger(CONTRACTS, 2);

        double barRange = high - low;
        double threshold = entryAtrMult * atr;

        if (barRange <= threshold) return;

        double stopDist = stopAtrMult * atr;
        if (stopDist <= 0) return;

        // Long: volatility expansion + bullish close
        if (close > open) {
            double stop = instr.round(close - stopDist);
            ctx.buy(contracts);
            entryPrice = close;
            stopPrice = stop;
            tradeSide = 1;
            bestPrice = close;
            tradesToday++;

            info(String.format("LONG: entry=%.2f, stop=%.2f, barRange=%.2f, ATR=%.2f (trail only, no TP)",
                close, stop, barRange, atr));

            drawEntryLevel(barTime, close, stop);

            var marker = settings.getMarker(Inputs.UP_MARKER);
            if (marker != null && marker.isEnabled()) {
                addFigure(new Marker(new Coordinate(barTime, low),
                    Enums.Position.BOTTOM, marker, "Long @ " + fmt(close)));
            }
        }
        // Short: volatility expansion + bearish close
        else if (close < open) {
            double stop = instr.round(close + stopDist);
            ctx.sell(contracts);
            entryPrice = close;
            stopPrice = stop;
            tradeSide = -1;
            bestPrice = close;
            tradesToday++;

            info(String.format("SHORT: entry=%.2f, stop=%.2f, barRange=%.2f, ATR=%.2f (trail only, no TP)",
                close, stop, barRange, atr));

            drawEntryLevel(barTime, close, stop);

            var marker = settings.getMarker(Inputs.DOWN_MARKER);
            if (marker != null && marker.isEnabled()) {
                addFigure(new Marker(new Coordinate(barTime, high),
                    Enums.Position.TOP, marker, "Short @ " + fmt(close)));
            }
        }
    }

    // ==================== Drawing ====================
    private static final java.awt.Color SL_COLOR = new java.awt.Color(220, 50, 50);
    private static final java.awt.Color ENTRY_COLOR = new java.awt.Color(100, 150, 255);
    private static final java.awt.Stroke DASH_STROKE = new java.awt.BasicStroke(
        1.5f, java.awt.BasicStroke.CAP_BUTT, java.awt.BasicStroke.JOIN_MITER,
        10.0f, new float[]{6.0f, 4.0f}, 0.0f);
    private static final java.awt.Font LEVEL_FONT = new java.awt.Font("SansSerif", java.awt.Font.PLAIN, 11);

    private void drawEntryLevel(long entryTime, double entry, double stop)
    {
        long endTime = entryTime + 4L * 60 * 60 * 1000;

        Line entryLine = new Line(entryTime, entry, endTime, entry);
        entryLine.setColor(ENTRY_COLOR);
        entryLine.setText("Entry " + fmt(entry), LEVEL_FONT);
        addFigure(entryLine);

        Line slLine = new Line(entryTime, stop, endTime, stop);
        slLine.setColor(SL_COLOR);
        slLine.setStroke(DASH_STROKE);
        slLine.setText("SL " + fmt(stop) + " (trail)", LEVEL_FONT);
        addFigure(slLine);
    }

    // ==================== Signal Handler ====================
    @Override
    public void onSignal(OrderContext ctx, Object signal)
    {
        // All trade logic handled in onBarUpdate()
    }

    // ==================== Utility ====================
    private int getTimeInt(long time, TimeZone tz)
    {
        Calendar cal = Calendar.getInstance(tz);
        cal.setTimeInMillis(time);
        return cal.get(Calendar.HOUR_OF_DAY) * 100 + cal.get(Calendar.MINUTE);
    }

    private int getDayOfYear(long time, TimeZone tz)
    {
        Calendar cal = Calendar.getInstance(tz);
        cal.setTimeInMillis(time);
        return cal.get(Calendar.DAY_OF_YEAR) + cal.get(Calendar.YEAR) * 1000;
    }

    private String fmt(double val) { return String.format("%.2f", val); }
}
